### PR TITLE
Waiting for maintenance mode flag

### PR DESCRIPTION
**Pull request by ebrahimradi**
@mateusjunges  This PR helps to wait for maintenance mode. I guess this is very helpful. what do you think?

### CODE REVIEW

File: src/Consumers/Consumer.php [hunk 1]
```diff
@@ -101,6 +101,10 @@ public function consume(): void
         }
 
         do {
+            while (app()->isDownForMaintenance()) {
+                $sleepTime = rand(1, 5);
```
Comment (mateusjunges) [inline src/Consumers/Consumer.php:105]: Why a random delay instead of a const?

**Review by mateusjunges — CHANGES_REQUESTED** [review]
I like this, but I think it would be better to offer the ability to also run a callback when the consumer detects that the application entered and left maintenance mode. We can define two callbacks that run in this two conditions. Something like this:


```php
if (! app()->isDownForMaintenance()) {
    return;
}

$enteringMaintenanceModeCallback = $this->whenEnterMaintenanceMode;
$enteringMaintenanceModeCallback(...)();

while (app()->isDownForMaintenance()) {
    $sleepTimeInSeconds = random_int(1, 5);
    sleep($sleepTimeInSeconds);
}

$exitingMaintenanceModeCallback = $this->whenExitingMaintenanceMode;
$exitingMaintenanceModeCallback(...)();
```

File: src/Consumers/Consumer.php (hunk 1 above)
Comment (ebrahimradi) [reply src/Consumers/Consumer.php:105]: not strong reason! just an example .

**Comment by ebrahimradi** [comment]
I wanted to apply your suggestion,

I findout a better, more generic idea.
instead of ONLY maintenance mode, we can have  a series of callbacks that can happen before consuming any message. something like middlewares but not related to messages.
then we can have lifecyle for consumer, before, and after. for now we can ignore after, but we can have before callbacks. what's your idea?

**Comment by mateusjunges** [comment]
Thanks! I have just merged #192 so I'll close this

**Pull request closed without merging**